applications/IgaApplication/custom_conditions/support_penalty_condition.cpp

```diff
@@ -0,0 +1,149 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Ricky Aristio
+//                   Tobias Teschemacher
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "custom_conditions/support_penalty_condition.h"
+
+
+namespace Kratos
+{
+    void SupportPenaltyCondition::CalculateAll(
+        MatrixType& rLeftHandSideMatrix,
+        VectorType& rRightHandSideVector,
+        ProcessInfo& rCurrentProcessInfo,
+        const bool CalculateStiffnessMatrixFlag,
+        const bool CalculateResidualVectorFlag
+    )
+    {
+        KRATOS_TRY
+        const double penalty = GetProperties()[PENALTY_FACTOR];
+
+        const auto& r_geometry = GetGeometry();
+        const SizeType number_of_nodes = r_geometry.size();
+        const SizeType mat_size = r_geometry.WorkingSpaceDimension() * number_of_nodes;
+
+        // Memory allocation
+        if (CalculateStiffnessMatrixFlag) {
+            if (rLeftHandSideMatrix.size1() != mat_size) {
+                rLeftHandSideMatrix.resize(mat_size, mat_size, false);
+            }
+            noalias(rLeftHandSideMatrix) = ZeroMatrix(mat_size, mat_size);
+        }
+        if (CalculateResidualVectorFlag) {
+            if (rRightHandSideVector.size() != mat_size) {
+                rRightHandSideVector.resize(mat_size, false);
+            }
+            rRightHandSideVector = ZeroVector(mat_size);
+        }
+
+
+        if (Has(DISPLACEMENT))
+        {
+            // Integration
+            const GeometryType::IntegrationPointsArrayType& integration_points = r_geometry.IntegrationPoints();
+            for (IndexType point_number = 0; point_number < integration_points.size(); point_number++)
+            {
+                Matrix N = r_geometry.ShapeFunctionsValues();
+
+                //FOR DISPLACEMENTS
+                Matrix H = ZeroMatrix(3, mat_size);
+                for (IndexType i = 0; i < number_of_nodes; i++)
+                {
+                    IndexType index = 3 * i;
+                    if (Is(IgaFlags::FIX_DISPLACEMENT_X))
+                        H(0, index) = N(point_number, i);
+                    if (Is(IgaFlags::FIX_DISPLACEMENT_Y))
+                        H(1, index + 1) = N(point_number, i);
+                    if (Is(IgaFlags::FIX_DISPLACEMENT_Z))
+                        H(2, index + 2) = N(point_number, i);
+                }
+
+                // Assembly
+                if (CalculateStiffnessMatrixFlag) {
+                    noalias(rLeftHandSideMatrix) += prod(trans(H), H)
+                        * penalty;
```

Comment:
* integration_points[i].Weight() * r_geometry.DeterminantOfJacobian()

Check the funciton in the CurveOnSurfaceQPG.